### PR TITLE
Use custom shared libraries from ffmpeg-build-scripts fork instead of homebrew

### PR DESCRIPTION
**Pull request by srubin**

### CODE REVIEW

File: binding.gyp [hunk 1]
```diff
@@ -35,8 +35,61 @@
...
+        "cflags_cc!": [
+          "-fno-rtti",
+          "-fno-exceptions"
+        ],
+        "cflags_cc": [
+          "-std=c++11",
+          "-fexceptions"
+        ],
+        "include_dirs": [
+          "<(module_root_dir)/ffmpeg/ffmpeg-ffprobe-shared-darwin-x86_64.1.21.rc4/include/"
```
Comment (srubin) [inline binding.gyp:62]: I'll find a way to not repeat this version number a bunch of places.

File: .github/workflows/prebuild.yml [hunk 2]
```diff
@@ -21,34 +21,43 @@ jobs:
...
       - name: install Linux dependencies
         run: |
           sudo add-apt-repository ppa:jonathonf/ffmpeg-4
           sudo apt-get update
           sudo apt-get install libavcodec-dev libavformat-dev libavdevice-dev libavfilter-dev libavutil-dev libpostproc-dev libswresample-dev libswscale-dev
         if: ${{ runner.os == 'Linux' }}
 
-      - name: install Mac dependencies
-        run: brew install nasm pkg-config ffmpeg
+      - name: install Mac dependencies (brew)
```
Comment (marcello3d) [inline .github/workflows/prebuild.yml:36]:
```suggestion
      - name: install Mac build dependencies (brew)
```

File: binding.gyp [hunk 3]
```diff
@@ -35,8 +35,61 @@
             "-lswresample",
             "-lswscale"
           ]
-        }
+        },
+        "copies": [
+            {
+              "destination": "build/Release/",
```
Comment (marcello3d) [inline binding.gyp:41]: wonder if we should use a nicer folder name like `lib` or `bin`

File: binding.gyp [hunk 4]
```diff
@@ -35,8 +35,61 @@
...
+          ],
+        },
+        'xcode_settings': {
+          'MACOSX_DEPLOYMENT_TARGET': '10.11',
+        },
+        "copies": [
+            {
+              "destination": "build/Release/",
+              "files": [
+                "<!@(node -p \"require('fs').readdirSync('ffmpeg/ffmpeg-ffprobe-shared-darwin-x86_64.1.21.rc4').map(f => 'ffmpeg/ffmpeg-ffprobe-shared-darwin-x86_64.1.21.rc4/' + f).join(' ')\")"
```
Comment (marcello3d) [inline binding.gyp:87]: this is kinda wow :D wonder if it should be in its own js file

File: patches/prebuild+10.0.1.patch [hunk 5]
```diff
@@ -0,0 +1,43 @@
+diff --git a/node_modules/prebuild/pack.js b/node_modules/prebuild/pack.js
+index 7552a4d..3be8aae 100644
+--- a/node_modules/prebuild/pack.js
++++ b/node_modules/prebuild/pack.js
+@@ -20,20 +20,30 @@ function pack (filenames, tarPath, cb) {
+     tarStream.pipe(zlib.createGzip({ level: 9 })).pipe(ws)
+ 
+     eachSeries(filenames, function processFile (filename, nextFile) {
+-      fs.stat(filename, function (err, st) {
++      fs.lstat(filename, function (err, st) {
```
Comment (marcello3d) [inline patches/prebuild+10.0.1.patch:10]: these changes feel reasonable to fork upstream

File: package.json [hunk 6]
```diff
@@ -47,11 +51,13 @@
   "homepage": "https://github.com/descriptinc/beamcoder#readme",
   "dependencies": {
     "bindings": "^1.5.0",
+    "ffmpeg-ffprobe-static": "^4.3.1-rc.2",
```
Comment (marcello3d) [inline package.json:54]: can this be a devDependency?

File: paths.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+const path = require('path');
+const os = require('os');
+
+function getPath(name) {
+  const binaries = Object.assign(Object.create(null), {
+    darwin: ['x64'],
+    linux: ['x64', 'ia32', 'arm64', 'arm'],
+    win32: ['x64', 'ia32']
+  });
```
Comment (marcello3d) [inline paths.js:30]: is this right?

File: install_ffmpeg.js [hunk 8]
```diff
@@ -175,35 +175,32 @@ sudo apt-get install libavcodec-dev libavformat-dev libavdevice-dev libavfilter-
...
+    else throw e;
+  });
 
-  return 0;
+  const ffmpegFilename = 'ffmpeg-ffprobe-shared-darwin-x86_64.1.21.rc4';
+  await access(`ffmpeg/${ffmpegFilename}`, fs.constants.R_OK).catch(async () => {
+    const ws = fs.createWriteStream(`ffmpeg/${ffmpegFilename}.zip`);
+    await get(
+      ws,
+      `https://github.com/descriptinc/ffmpeg-build-script/releases/download/v1.21.rc4/${ffmpegFilename}.zip`,
```
Comment (marcello3d) [inline install_ffmpeg.js:190]: wonder if these versions should be pulled up to the top somewhere

File: .github/workflows/prebuild.yml [hunk 9]
```diff
@@ -21,34 +21,43 @@ jobs:
...
-        run: npx prebuild --runtime napi --arch arm64 --prepack copy_mac_libs.js --include-regex '\.(node|dylib)$' --upload ${{ secrets.GITHUB_TOKEN }}
-        if: ${{ runner.os == 'macOS'}}
+#      - name: prebuild for macOS (arm64)
+#        run: npx prebuild --runtime napi --arch arm64 --include-regex '(\.node|\.dylib|ffmpeg|ffprobe)$' --upload ${{ secrets.GITHUB_TOKEN }}
+#        if: ${{ runner.os == 'macOS'}}
 
       - name: prebuild for Windows
-        # I can't get the pipe character to work on windows, so to include both .node and .dll files I have this weird regexp
-        run: npx prebuild --runtime napi --include-regex '\.[nd][ol][dl]e?$' --upload ${{ secrets.GITHUB_TOKEN }}
+        run: npx prebuild --runtime napi --include-regex '"(.node|.dll|ffmpeg.exe|ffprobe.exe)$"' --upload ${{ secrets.GITHUB_TOKEN }}
```
Comment (marcello3d) [inline .github/workflows/prebuild.yml:57]: were you able to get this to work? I couldn't get `|` to work on windows for some reason… also I think you need to escape the `.`:
```suggestion
        run: npx prebuild --runtime napi --include-regex '"(\.node|\.dll|ffmpeg\.exe|ffprobe\.exe)$"' --upload ${{ secrets.GITHUB_TOKEN }}
```

**Review by marcello3d — APPROVED** [review]

File: package.json (hunk 6 above)
Comment (srubin) [reply package.json:54]: Yes

File: .github/workflows/prebuild.yml (hunk 9 above)
Comment (srubin) [reply .github/workflows/prebuild.yml:57]: Yes, I got it to work by switching to the weird single and double quotes and switching to git-bash for the shell.

**Pull request merged**